openedx/features/survey_report/queries.py

```diff
@@ -0,0 +1,58 @@
+"""
+Queries to get data from database.
+"""
+
+from datetime import datetime, timedelta
+
+from django.contrib.auth.models import User  # lint-amnesty, pylint: disable=imported-auth-user
+from django.db.models import Count, OuterRef, Q, Subquery
+
+from common.djangoapps.student.models import CourseEnrollment
+from lms.djangoapps.grades.models import PersistentCourseGrade
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+
+
+def get_unique_courses_offered() -> int:
+    """
+    Get total number of unique courses offered.
+    """
+    return CourseOverview.objects.annotate(
+        count=Subquery(
+            CourseEnrollment.objects
+                            .filter(course_id=OuterRef('id'))
+                            .values('course_id')
+                            .annotate(count=Count('course_id'))
+                            .values('count')
+        ))\
+        .filter(count__gt=5)\
+        .filter(start__lt=datetime.now())\
+        .filter(Q(end__isnull=True) | Q(end__gt=datetime.now()))\
+        .count()
+
+
+def get_currently_learners() -> int:
+    """
+    Get total number of learners with last login in the last 3 weeks.
+    """
+    return User.objects.filter(last_login__gte=datetime.now() - timedelta(weeks=3)).count()
+
+
+def get_learners_registered() -> int:
+    """
+    Get total number of active learners registered.
+    """
+    return User.objects.filter(is_active=True).count()
+
+
+def get_generated_certificates() -> int:
+    """
+    Get total number of generated certificates.
+    """
+    return PersistentCourseGrade.objects.filter(passed_timestamp__isnull=False).count()
```

Comment:
Is this the right table? Don't people who have persistent grades disabled still have certificates generated in a separate certificates model?

Reply:
I think this a mix of confusing naming with a feature that requires a fallback.

In the [first survey queries](https://gist.github.com/feanil/0b49d5219ae08baf1216556d455dfdbb) that were published in the [blog](https://openedx.org/blog/open-edx-survey-help-us-quantify-the-impact-of-open-edx-globally/) we saying that this would render the total number of course completions. Also we made sure to note that this would only work if you had enabled persistent grades ([recently](https://github.com/openedx/edx-platform/pull/29793) certificate generation was not even working without the enabled feature).

If we wanted the certificates generated I would argue we should query the `GeneratedCertificates` model, but not every course will configure certificates and that would be required for the actual generation to happen.

Now, according to [this comment](https://github.com/openedx/edx-platform/pull/31183#issuecomment-1295177113) what we want is a number that represents "How many learners have completed an Open edX course?". Which is a whole question on itself since there is nothing that marks the end of a course in a conclusive way, but to this end PersistentCourseGrade is probably a better proxy.

My proposal would be to call this method `get_completed_courses` check if persistent grades is enabled and if yes, use the `PersistentCourseGrade`. Otherwise use the `GeneratedCertificates`. In any case we leave a proper docstring explaining the difficult situation.

Reply:
In case we are interested in the course completion the number of earned certificates really doesn't work in many cases. The persistent grades are also not covered all the cases since there could be courses without a lot of or even any graded assignments. I'm not sure how big or small the part of the courses without graded assignments but for that part, we could consider looking at the `BlockCompletion` model if the completion feature is enabled for the course.

Reply:
You are actually getting to the core of the issue. There is not one way of deciding that a course has been completed. Some courses use the certificate as the marker for finishing, some others require a certain grade. Both of those counts will be a little different.

Using BlockCompletion would also be a possibility, but since we are trying to get the completed courses, that would require us to come up with a new metric to map the number of "block completions" to a "completed course".
If I were to give it a shot I would say after a certain % of completed blocks against the total of blocks then the course is completed. This % should be configurable by the instructors. This is however way out of the scope of this PR. That is why we are taking the original set of queries from the survey sent to the community before the last conference and we are automating that.

Reply:
I'll resolve for now and continue to merging since using blockCompletion for the metric would be additive work and does not need to block the current PR.